mem.go

```diff
@@ -571,40 +573,51 @@ func (e *Encoder) Encode(m *Message) error {
 	if nsegs == 0 {
 		return errMessageEmpty
 	}
-	sizes, err := m.segmentSizes()
-	if err != nil {
-		return err
-	}
-	maxSeg := uint32(nsegs - 1)
-	hdrSize := streamHeaderSize(maxSeg)
-	if uint64(cap(e.hdrbuf)) < hdrSize {
-		e.hdrbuf = make([]byte, hdrSize)
+	if int64(cap(e.bufs)) < 1+nsegs {
+		e.bufs = make([][]byte, 1+nsegs)
 	} else {
-		e.hdrbuf = e.hdrbuf[:hdrSize]
+		e.bufs = e.bufs[:1+nsegs]
 	}
-	marshalStreamHeader(e.hdrbuf, sizes)
-	if err := e.write(e.hdrbuf); err != nil {
-		return err
+	if int64(cap(e.sizes)) < nsegs {
+		e.sizes = make([]Size, nsegs)
+	} else {
+		e.sizes = e.sizes[:nsegs]
 	}
 	for i := int64(0); i < nsegs; i++ {
 		s, err := m.Segment(SegmentID(i))
 		if err != nil {
 			return err
 		}
-		if err := e.write(s.data); err != nil {
-			return err
+		n := len(s.data)
+		if int64(n) > int64(maxSize) {
+			return errSegmentTooLarge
 		}
+		e.sizes[i] = Size(n)
+		e.bufs[1+i] = s.data
 	}
-	return nil
+	maxSeg := uint32(nsegs - 1)
+	hdrSize := streamHeaderSize(maxSeg)
+	if uint64(cap(e.hdrbuf)) < hdrSize {
+		e.hdrbuf = make([]byte, hdrSize)
+	} else {
+		e.hdrbuf = e.hdrbuf[:hdrSize]
+	}
+	marshalStreamHeader(e.hdrbuf, e.sizes)
```

Comment:
I've only touched the `(*Encoder).Encode` part so far and I've left the use of `marshalStreamHeader` in `(*Message).Marshal`. So the bug still exists in `(*Message).Marshal`.

I'm not sure how you want me to handle that, but I'm thinking it would be best if `(*Message).Marshal` and `(*Message).MarshalPacked` were just made wrappers around `(*Encoder).Encode`. That would mean the encoding logic only exists once and would reduce duplication. I think I'll go ahead and add a commit for that and you can decide whether you like that method or not.